source/material_model/reaction_model/katz2003_mantle_melting.cc

```diff
@@ -150,6 +168,203 @@ namespace aspect
+                                              * out.thermal_expansion_coefficients[i];
+                  }
+                else
+                  temperature_dependence -= (in.temperature[i] - reference_T) * out.thermal_expansion_coefficients[i];
+
+                // the fluid compressibility includes two parts, a constant compressibility, and a pressure-dependent one
+                // this is a simplified formulation, experimental data are often fit to the Birch-Murnaghan equation of state
+                const double fluid_compressibility = melt_compressibility / (1.0 + in.pressure[i] * melt_bulk_modulus_derivative * melt_compressibility);
+
+                melt_out->fluid_densities[i] = reference_rho_fluid * std::exp(fluid_compressibility * (in.pressure[i] - this->get_surface_pressure()))
+                                               * temperature_dependence;
+
+                melt_out->fluid_density_gradients[i] = melt_out->fluid_densities[i] * melt_out->fluid_densities[i]
+                                                       * fluid_compressibility
+                                                       * this->get_gravity_model().gravity_vector(in.position[i]);
+
+                const double phi_0 = 0.05;
+                porosity = std::max(std::min(porosity,0.995),1e-4);
+                melt_out->compaction_viscosities[i] = xi_0 * phi_0 / porosity;
+
+                double visc_temperature_dependence = 1.0;
+                if (this->include_adiabatic_heating ())
+                  {
+                    const double delta_temp = in.temperature[i]-this->get_adiabatic_conditions().temperature(in.position[i]);
+                    visc_temperature_dependence = std::max(std::min(std::exp(-thermal_bulk_viscosity_exponent*delta_temp/this->get_adiabatic_conditions().temperature(in.position[i])),1e4),1e-4);
+                  }
+                else
+                  {
+                    const double delta_temp = in.temperature[i]-reference_T;
+                    const double T_dependence = (thermal_bulk_viscosity_exponent == 0.0
+                                                 ?
+                                                 0.0
+                                                 :
+                                                 thermal_bulk_viscosity_exponent*delta_temp/reference_T);
+                    visc_temperature_dependence = std::max(std::min(std::exp(-T_dependence),1e4),1e-4);
+                  }
+                melt_out->compaction_viscosities[i] *= visc_temperature_dependence;
+
+                if (this->include_melt_transport() && in.requests_property(MaterialProperties::viscosity))
+                  {
+                    const double porosity = std::min(1.0, std::max(in.composition[i][porosity_idx],0.0));
+                    out.viscosities[i] *= exp(- alpha_phi * porosity);
+                  }
```

Comment:
This should be outside of the loop `if (melt_out != nullptr)`. 
Otherwise the weakening due to the presence of melt depends on whoever calls this function having created the melt_outputs.